src/crystal/system/win32/file.cr

```diff
@@ -96,6 +97,24 @@ module Crystal::System::File
     {access, disposition, attributes}
   end
 
+  protected def system_set_mode(mode : String)
+    @system_append = true if mode.starts_with?('a')
+  end
+
+  private def write_blocking(handle, slice)
+    if @system_append
+      write_blocking(handle, slice) do
+        overlapped = LibC::OVERLAPPED.new
+        overlapped.union.offset.offset = 0xFFFFFFFF_u32
+        overlapped.union.offset.offsetHigh = 0xFFFFFFFF_u32
+        ret = LibC.WriteFile(handle, slice, slice.size, out bytes_written, pointerof(overlapped))
+        {ret, bytes_written}
+      end
+    else
+      super
+    end
+  end
```

Comment:
We could also consider moving `@system_append` to `Crystal::System::FileDescriptor` because that's where it matters. Might also make sense to expose it in `FileDescriptor.new` to use this behaviour with an existing handle.

Reply:
That snippet wouldn't work as `pointerof(overlapped)` should not be a `Pointer(LibC::OVERLAPPED | Nil)`

Reply:
You're probably right that `IO::FileDescriptor` could open a regular file via its `#fd` directly and it would lose the append mode. I'm not sure if there is a way around that

Reply:
Sorry, I scribbled the code down too quickly. `pointerof(overlapped)` should go in the if branch, of course.

Reply:
> You're probably right that `IO::FileDescriptor` could open a regular file via its `#fd` directly and it would lose the append mode. I'm not sure if there is a way around that

There's probably no way around it. We could try to be smart and query the access mode of the file descriptor to figure out the intention. But we cannot change it anyway, so it's probably a mute point.
It's probably fine to accept that an external file descriptor without `FILE_WRITE_DATA` access mode won't be able to truncate or lock.
But we can offer an option for a file descriptor *with* `FILE_WRITE_DATA` to only append to the end of the file if we allow to configure `@system_access` in the constructor.